Cryptic compilation error with Spock 2.0-M1 and Groovy 3.0.0-RC2

A try to Spock 2.0-M1 with Groovy 3.0.0-RC2 (manually playing with Groovy dependencies) results in very cryptic error message:

```
> Task :compileTestGroovy FAILED
startup failed:
Could not instantiate global transform class org.spockframework.compiler.SpockTransform specified at
jar:file:/.../spock-core-2.0-M1-groovy-2.5.jar!/META-INF/services/org.codehaus.groovy.transform.ASTTransformation
because of exception java.lang.reflect.InvocationTargetException

1 error
```

Having Groovy 2.4 and Spock 1.3-groovy-2.5 returns much more verbose:

```
> Task :compileTestGroovy FAILED
startup failed:
Could not instantiate global transform class org.spockframework.compiler.SpockTransform specified at
jar:file:/.../spock-core-1.3-groovy-2.5.jar!/META-INF/services/org.codehaus.groovy.transform.ASTTransformation
because of exception org.spockframework.util.IncompatibleGroovyVersionException:
The Spock compiler plugin cannot execute because Spock 1.3.0-groovy-2.5 is not compatible
with Groovy 2.4.3. For more information, see http://versioninfo.spockframework.org
Spock artifact: file:/.../spock-core-1.3-groovy-2.5.jar
Groovy artifact: file:/.../groovy-2.4.3.jar

1 error
```

It would be good to try to detect that incompatibility earlier and fail with more verbose error message for Spock 2.0 and Groovy 3.0.

Btw, as at least basic tests pass with Spock 2.0-groovy-2.5-SNAPSHOT and Groovy 3.0.0-RC2 maybe it would be good to allow to run Spock 2.0-M2 with Groovy 3.x to facilitate testing that configuration (before Spock can be built and tested with Groovy 3)?

The same occurrs with M2 and Groovy 3.0.1:

 

```

Could not instantiate global transform class org.spockframework.compiler.SpockTransform specified at jar:file:/d:/m2_repo/org/spockframework/spock-core/2.0-M2-groovy-3.0/spock-core-2.0-M2-groovy-3.0.jar!/META-INF/services/org.codehaus.groovy.transform.ASTTransformation  because of exception java.lang.reflect.InvocationTargetException

```

 

JDK12, gmavenplus-plugin:1.8.1:compileTests

 

it seems M2 is unusable with Groovy 3 for now

 

Full trace:

 

```

org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.gmavenplus:gmavenplus-plugin:1.8.1:compileTests (default) on project common: Error occurred while calling a method on a Groovy class from classpath.

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)

    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)

    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)

    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)

    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)

    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)

    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)

    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)

    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)

    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)

    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)

    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)

    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)

    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)

    at java.lang.reflect.Method.invoke (Method.java:566)

    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)

    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)

    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)

    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)

    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)

Caused by: org.apache.maven.plugin.MojoExecutionException: Error occurred while calling a method on a Groovy class from classpath.

    at org.codehaus.gmavenplus.mojo.CompileTestsMojo.execute (CompileTestsMojo.java:80)

    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)

    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)

    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)

    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)

    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)

    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)

    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)

    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)

    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)

    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)

    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)

    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)

    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)

    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)

    at java.lang.reflect.Method.invoke (Method.java:566)

    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)

    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)

    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)

    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)

    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)

Caused by: java.lang.reflect.InvocationTargetException

    at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0 (Native Method)

    at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance (NativeConstructorAccessorImpl.java:62)

    at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance (DelegatingConstructorAccessorImpl.java:45)

    at java.lang.reflect.Constructor.newInstance (Constructor.java:490)

    at org.codehaus.gmavenplus.util.ReflectionUtils.invokeConstructor (ReflectionUtils.java:191)

    at org.codehaus.gmavenplus.mojo.AbstractCompileMojo.setupCompilationUnit (AbstractCompileMojo.java:291)

    at org.codehaus.gmavenplus.mojo.AbstractCompileMojo.doCompile (AbstractCompileMojo.java:263)

    at org.codehaus.gmavenplus.mojo.CompileTestsMojo.execute (CompileTestsMojo.java:76)

    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)

    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)

    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)

    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)

    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)

    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)

    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)

    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)

    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)

    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)

    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)

    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)

    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)

    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)

    at java.lang.reflect.Method.invoke (Method.java:566)

    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)

    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)

    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)

    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)

    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)

Caused by: org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:

Could not instantiate global transform class org.spockframework.compiler.SpockTransform specified at jar:file:/d:/m2_repo/org/spockframework/spock-core/2.0-M2-groovy-3.0/spock-core-2.0-M2-groovy-3.0.jar!/META-INF/services/org.codehaus.groovy.transform.ASTTransformation  because of exception java.lang.reflect.InvocationTargetException

 

1 error

 

    at org.codehaus.groovy.control.ErrorCollector.failIfErrors (ErrorCollector.java:311)

    at org.codehaus.groovy.control.ErrorCollector.addError (ErrorCollector.java:108)

    at org.codehaus.groovy.transform.ASTTransformationVisitor.addPhaseOperationsForGlobalTransforms (ASTTransformationVisitor.java:333)

    at org.codehaus.groovy.transform.ASTTransformationVisitor.doAddGlobalTransforms (ASTTransformationVisitor.java:280)

    at org.codehaus.groovy.transform.ASTTransformationVisitor.addGlobalTransforms (ASTTransformationVisitor.java:204)

    at org.codehaus.groovy.transform.ASTTransformationVisitor.addPhaseOperations (ASTTransformationVisitor.java:168)

    at org.codehaus.groovy.control.CompilationUnit.addPhaseOperations (CompilationUnit.java:245)

    at org.codehaus.groovy.control.CompilationUnit.<init> (CompilationUnit.java:175)

    at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0 (Native Method)

    at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance (NativeConstructorAccessorImpl.java:62)

    at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance (DelegatingConstructorAccessorImpl.java:45)

    at java.lang.reflect.Constructor.newInstance (Constructor.java:490)

    at org.codehaus.gmavenplus.util.ReflectionUtils.invokeConstructor (ReflectionUtils.java:191)

    at org.codehaus.gmavenplus.mojo.AbstractCompileMojo.setupCompilationUnit (AbstractCompileMojo.java:291)

    at org.codehaus.gmavenplus.mojo.AbstractCompileMojo.doCompile (AbstractCompileMojo.java:263)

    at org.codehaus.gmavenplus.mojo.CompileTestsMojo.execute (CompileTestsMojo.java:76)

    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)

    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)

    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)

    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)

    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)

    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)

    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)

    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)

    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)

    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)

    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)

    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)

    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)

    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)

    at java.lang.reflect.Method.invoke (Method.java:566)

    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)

    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)

    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)

    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)

    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)

```

--- a/build.gradle
+++ b/build.gradle
@@ -21,7 +21,7 @@ ext {
   variants = [2.5, 3.0]
   variant = System.getProperty("variant") as BigDecimal ?: variants.first()
   if (variant == 2.5) {
-      groovyVersion = "2.5.8"
+      groovyVersion = "2.5.11"
       minGroovyVersion = "2.5.0"
       maxGroovyVersion = "2.9.99"
     } else if (variant == 3.0) {

--- a/build.gradle
+++ b/build.gradle
@@ -25,7 +25,7 @@ ext {
       minGroovyVersion = "2.5.0"
       maxGroovyVersion = "2.9.99"
     } else if (variant == 3.0) {
-      groovyVersion = "3.0.2"
+      groovyVersion = "3.0.3"
       minGroovyVersion = "3.0.0"
       maxGroovyVersion = "3.9.99"
     } else {

--- a/docs/release_notes.adoc
+++ b/docs/release_notes.adoc
@@ -188,6 +188,8 @@ Spy(constructorArgs: [null as String, (Pattern) null])
 def "I'll run everywhere but on Windows"() { ... }
 ----
 
+- Upgrade Groovy to 2.5.11 (improved Java 14+ support) and 3.0.3 (fixes #1127)
+
 - Upgrade JUnit 4 to 4.13 (in `spock-junit4`)
 
 - Upgrade Hamcrest to 2.2 (from 1.3 provided previously by `junit4.jar`)

--- a/spock-junit4/src/test/groovy/org/spockframework/smoke/groovy/UsageOfDeprecatedNotYetImplemented.groovy
+++ b/spock-junit4/src/test/groovy/org/spockframework/smoke/groovy/UsageOfDeprecatedNotYetImplemented.groovy
@@ -17,17 +17,14 @@ package org.spockframework.smoke.groovy
 
 import groovy.transform.NotYetImplemented
 import junit.framework.AssertionFailedError
+import spock.lang.Issue
 import spock.lang.Specification
 import spock.lang.FailsWith
 
-//TODO: Handle also new groovy.transform.NotYetImplemented in Groovy 3 - https://github.com/spockframework/spock/issues/1127
-class UsageOfNotYetImplemented extends Specification {
-
-  @FailsWith(AssertionFailedError)
-  @NotYetImplemented
-  def "not allowed to pass"() {
-    expect: true
-  }
+//See UsageOfNotYetImplemented for new groovy.test.NotYetImplemented tests
+@Issue("https://github.com/spockframework/spock/issues/1127")
+@Deprecated
+class UsageOfDeprecatedNotYetImplemented extends Specification {
 
   @NotYetImplemented
   def "expected to fail (legacy transformation)"() {

--- a/spock-specs/specs.gradle
+++ b/spock-specs/specs.gradle
@@ -11,6 +11,9 @@ configurations {
 dependencies {
   testCompile project(":spock-core")
   testCompile libs.groovySql  //for groovy.sql.Sql
+  testCompile(libs.groovyTest) {  //for groovy.test.NotYetImplemented
+    exclude group: 'junit'
+  }
 
   testRuntime libs.asm
   testRuntime libs.bytebuddy

--- a/spock-specs/src/test3.0/groovy/org/spockframework/smoke/UsageOfNotYetImplemented.groovy
+++ b/spock-specs/src/test3.0/groovy/org/spockframework/smoke/UsageOfNotYetImplemented.groovy
@@ -1,0 +1,38 @@
+package org.spockframework.smoke
+
+import groovy.test.NotYetImplemented
+import org.opentest4j.AssertionFailedError
+import org.spockframework.runtime.GroovyRuntimeUtil
+import spock.lang.FailsWith
+import spock.lang.Issue
+import spock.lang.Requires
+import spock.lang.Specification
+
+//Those tests requires Groovy 3.0.3+, see UsageOfNotYetImplementedJUnit4 for deprecated groovy.transform.NotYetImplemented tests
+@Issue(["https://github.com/spockframework/spock/issues/1127", "https://issues.apache.org/jira/browse/GROOVY-9492"])
+@Requires({ GroovyRuntimeUtil.isGroovy3orNewer() })
+class UsageOfNotYetImplemented extends Specification {
+
+  @NotYetImplemented
+  def "expected to fail"() {
+    expect: false
+  }
+
+  @NotYetImplemented
+  def "allowed to raise arbitrary exception"() {
+    setup:
+    throw new IOException("ouch")
+  }
+
+  @FailsWith(AssertionError)
+  @NotYetImplemented
+  def "not allowed to pass"() {
+    expect: true
+  }
+
+  @FailsWith(AssertionFailedError)
+  @NotYetImplemented(exception = AssertionFailedError)
+  def "not allowed to pass with custom exception type"() {
+    expect: true
+  }
+}

